iree/turbine/kernel/wave/codegen/read_write.py

```diff
@@ -583,28 +641,67 @@ def _create_vec_write(
         data, offset_th = _linearize_memref(
             mem, start_indices_wg, start_indices_th, strides
         )
-        offset_th = vector_d.splat(offsets_vec.type, offset_th)
-        offsets_vec = arith_d.addi(offsets_vec, offset_th)
-        if mask is not None:
+
+        if splatted_masked:
+            # If mask value is same for all vector elements, we can use vector
+            # buffer ops.
             i32 = IntegerType.get_signless(32)
-            i32vec = VectorType.get([elements_per_thread], i32)
-            offsets_vec = arith_d.index_cast(i32vec, offsets_vec)
+            offset_th = arith_d.index_cast(i32, offset_th)
             oob_idx = _get_max_buffer_size(element_type)
             oob_idx = arith_d.constant(i32, oob_idx)
-            oob_idx = vector_d.splat(offsets_vec.type, oob_idx)
-            offsets_vec = arith_d.select(mask, offsets_vec, oob_idx)
+            offset_th = arith_d.select(mask_splat, offset_th, oob_idx)
 
-        for i in range(elements_per_thread):
-            offset = vector_d.extract(
-                offsets_vec, static_position=[i], dynamic_position=[]
-            )
-            elem = vector_d.extract(value, static_position=[i], dynamic_position=[])
-            if mask is None:
-                memref_d.store(elem, data, indices=[offset])
-            else:
-                amdgpu_d.raw_buffer_store(elem, data, indices=[offset])
+            # Buffer ops doesn't support 1-element vectors, convert to scalar.

```

Comment:
Is this an LLVM/MLIR bug? If so can you mark it as a TODO for someone to upstream the fix?

Reply:
Upstream fIx is merged https://github.com/llvm/llvm-project/pull/128266, we need to wait until IREE update

Reply:
nice